File: dev-js-api/js/algorithms/placement/departments.js
/**
 * @fileoverview departments.js — Pure algorithms for department placement and AABB computations.
 */

import { packRectangles } from './shards.js';

/**
 * Packs departments relative to each other on a level.
 * 
 * @param {Array} deptRects - Department rectangles to pack
 * @param {number} gapDepts - Gap spacing between departments
 * @returns {Object<string, {u: number, v: number}>} Relative positions of departments
 */
export function packDepartmentsOnLevel(deptRects, gapDepts) {
  const { positions } = packRectangles(deptRects, gapDepts);
  return positions;
}

/**
 * Dynamically computes AABB boundaries for departments based on member shards' coordinates.
 * 
 * @param {Array} shardsOut - Placed shards list with absolute coordinates
 * @returns {Array} List of resolved departments with position and size
 */
export function computeDepartmentBounds(shardsOut) {
  const departmentsOut = [];
  const resolvedDepts = {};

  shardsOut.forEach(s => {
    const lvlId = s.orbit;
    if (!resolvedDepts[s.dept]) {
      resolvedDepts[s.dept] = {
        name: s.dept,
        orbit: lvlId,
        x_min: s.position.x,
        x_max: s.position.x + s.size.w,
        y_min: s.position.y,
        y_max: s.position.y + s.size.d
      };
    } else {
      const dObj = resolvedDepts[s.dept];
      dObj.x_min = Math.min(dObj.x_min, s.position.x);
      dObj.x_max = Math.max(dObj.x_max, s.position.x + s.size.w);
      dObj.y_min = Math.min(dObj.y_min, s.position.y);
      dObj.y_max = Math.max(dObj.y_max, s.position.y + s.size.d);
    }
  });

  Object.values(resolvedDepts).forEach(d => {
    departmentsOut.push({
      name: d.name,
      orbit: d.orbit,
      position: { x: d.x_min, y: d.y_min },
      size: { w: d.x_max - d.x_min, d: d.y_max - d.y_min }
    });
  });

  return departmentsOut;
}

File: dev-client-api/algorithms/placement/departments.py
#!/usr/bin/env python3
"""
departments.py — Pure Python functions for department placement and AABB bounds calculation.
"""

from algorithms.placement.shards import pack_rectangles

def pack_departments_on_level(dept_rects, gap_depts):
    """
    Packs department rectangles relative to each other.
    Returns: positions (dict of dept_name -> (u, v))
    """
    _, _, positions = pack_rectangles(dept_rects, gap_depts)
    return positions


def compute_department_bounds(shards_out):
    """
    Calculates dynamic AABB boundaries for departments based on final shard coordinates.
    Returns: list of dicts representing departments with position and size
    """
    departments_out = []
    resolved_depts = {}

    for s in shards_out:
        lvl_id = s["orbit"]
        dname = s["dept"]
        
        px = s["position"]["x"]
        py = s["position"]["y"]
        sw = s["size"]["w"]
        sd = s["size"]["d"]

        if dname not in resolved_depts:
            resolved_depts[dname] = {
                "name": dname,
                "orbit": lvl_id,
                "x_min": px,
                "x_max": px + sw,
                "y_min": py,
                "y_max": py + sd
            }
        else:
            d_obj = resolved_depts[dname]
            d_obj["x_min"] = min(d_obj["x_min"], px)
            d_obj["x_max"] = max(d_obj["x_max"], px + sw)
            d_obj["y_min"] = min(d_obj["y_min"], py)
            d_obj["y_max"] = max(d_obj["y_max"], py + sd)

    for d in resolved_depts.values():
        departments_out.append({
            "name": d["name"],
            "orbit": d["orbit"],
            "position": {"x": d["x_min"], "y": d["y_min"]},
            "size": {"w": d["x_max"] - d["x_min"], "d": d["y_max"] - d["y_min"]}
        })

    return departments_out
